runtime/lua/vim/treesitter/dev.lua

```diff
@@ -436,11 +457,150 @@ function M.inspect_tree(opts)
     buffer = buf,
     once = true,
     callback = function()
-      if api.nvim_win_is_valid(w) then
-        api.nvim_win_close(w, true)
+      close_win(w)
+    end,
+  })
+end
+
+-- TODO(MariaSolOs): Add helper functions for common code
+-- between the query editor and the tree inspector.
+--- @class EditQueryOpts
+--- @field bufnr integer? Buffer to draw the editor into. If omitted, a new
+---                       buffer is created.
+--- @field winid integer? Window id to display the editor buffer in. If omitted,
+---                       a new window is created with {command}.
+--- @field command string? Vimscript command to create the window. Default
+---                        value is "60vnew" or "new" if the inspector is open.
+---                        Only used when {winid} is nil.
+
+--- @private
+--- @param opts EditQueryOpts?
+M.edit_query = function(opts)
+  vim.validate({
+    opts = { opts, 't', true },
+  })
+
+  opts = opts or {}
+
+  local buf = api.nvim_get_current_buf()
+
+  -- Close any existing editor window
+  if vim.b[buf].dev_edit then
+    close_win(vim.b[buf].dev_edit)
+  end
+
+  local w = opts.winid
+  if not w then
+    local cmd = opts.command or '60vnew'
+    -- If the inspector is open, place the editor above it.
+    local base_buf = vim.b[buf].dev_base ---@type integer?
+    local inspect_buf = base_buf and vim.b[base_buf].dev_inspect
+    if base_buf and api.nvim_win_is_valid(inspect_buf) then
+      vim.api.nvim_set_current_win(inspect_buf)
+      buf = base_buf
+      cmd = 'new'
+    end
+    vim.cmd(cmd)
+    w = api.nvim_get_current_win()
+  end
+
+  local b = opts.bufnr
+  if b then
+    api.nvim_win_set_buf(w, b)
+  else
+    b = api.nvim_win_get_buf(w)
+  end
+
+  vim.b[buf].dev_edit = w
+  vim.bo[b].omnifunc = 'v:lua.vim.treesitter.query.omnifunc'
```

Comment:
Suggestion: add an `opts` field with `lang` to `query.omnifunc` that is used instead of `guess_lang` if set, and create a closure here to set `vim.bo.omnifunc` to.

(Could do the same in the query ftplugin; that'd save the constant lookup.)

Reply:
Thanks for the hint @clason! I'm having some trouble creating such closure though. I've tried:
```lua
local omnifunc_opts = { lang = lang }
vim.bo[b].omnifunc =
  '{findstart, base -> v:lua.vim.treesitter.query.omnifunc(findstart, base, omnifunc_opts)}'

-- Inside M.edit_query
M.omnifunc = function(findstart, base)
  return vim.treesitter.query.omnifunc(findstart, base, { lang = lang })
end
vim.bo[b].omnifunc = 'v:lua.vim.treesitter.dev.omnifunc'
```

Is this the right syntax?

Reply:
Alternatively I can just add a buffer property to the query editor buffer, do not introduce a new parameter in `omnifunc`, and instead use that new buffer property as a fallback if `guess_query_lang` fails.

Reply:
Okay I came up with another hack. Let me know what you think :)

Reply:
No, it's fine, your trick is good for now (at least). We can still revisit if it turns out there are problems with it -- and if not, it clearly works fine ;)

Reply:
Sounds good :)